### DISCUSSION TITLE
`exclude_none` for specific fields only

### DISCUSSION THREAD

**Discussion by mar-muel** [pydantic/pydantic#5461] [Feature Request]
Hi there! Apologies for asking stuff that is probably trivial, but couldn't find an answer to this.

I know it's possible to exclude `None` values globally. In my case, I'm generating a JSON response from FastAPI with Pydantic, and I would like to exclude only certain keys if None, but for all other fields, keep the default to showing null values, as sometimes they are meaningful.

What would be the easiest way to achieve this? Thanks for your help 🙏

**Comment by dmontagu** [comment]
In v2 this could be done with a `model_serializer`.

@samuelcolvin @adriangb is there any way we could make it possible for a `field_serializer` to _exclude_ the field? Like, make it so the value doesn't end up in the final serialized object at all? I'm thinking the API could be something like raising a special exception in the field_serializer (`raise pydantic.DoNotSerialize`). That would prevent the need to handle certain return types specially, but maybe it's too weird. Or maybe it wouldn't work well with pydantic_core.

Either way, I think it would be somewhat powerful if we could support this using a field serializer, as it would allow you to get the exclude-if-none behavior through a type annotation.

**Reply by adriangb** [reply]
Might be better if we had an `SkipSer[T]` annotation then that is only legal in model fields?

```python
class Model(BaseModel):
  x: int
  y: int
  description: SkipSer[int]
```

Then if we encounter that in a model field we can set something in the core schema that excludes it?

**Reply by adriangb** [reply]
Ah but that doesn't help with:
- Skipping only if the value is the default value ('ve wanted to do this in the past)
- Skip if the value is a specific value (e.g. `None`)

**Reply by dmontagu** [reply]
Yeah, I think the big win comes from being able to use arbitrary logic to determine whether to include the field — it would mean that _any_ use case a user might want can be handled by their own logic, rather than us having a bunch of types/flags/etc to maintain. (Especially if the `info.context` thing becomes a reality, in which case they could control it on a per-call basis.)

**Reply by mar-muel** [reply]
I fully agree with the need to be able to provide custom logic. Just as an example: If I'm using `orm` mode then I rely on logic that certain fields are not eagerly loaded. So depending on how you implement these flags `Model.field is None` (rather than something like `inspect(self._obj).attrs.field.loaded_value == NO_VALUE`) would trigger an SQL sub query, which is probably not wanted by the user.

**Reply by adriangb** [reply]
@dmontagu I think we could make something like this work:

```python
from __future__ import annotations

from typing import Any, TypeVar

from pydantic_core import PydanticOmit
from typing_extensions import Annotated

import pydantic


def omit_if_none(v: Any, handler: pydantic.SerializerFunctionWrapHandler) -> str | None:
    # careful with omitting list items, see note below
    #   - raising a PydanticOmit in a list should exclude that item from the list not the list from the model...
    #     do we want a way to exclude the field based on a serializer of a list item? I don't think so...
    #   - ideally the same function can be used for both cases and differentiated them based on the info param
    #     that is: the function shouldn't create a bug where it omits list items when it meant to omit model fields
    if v is None:
        raise PydanticOmit
    return handler(v)


def omit_if_default(v: Any, handler: pydantic.SerializerFunctionWrapHandler, info: pydantic.SerializationInfo) -> str | None:
    # need:
    # - current field name
    # - know which fields were set and which were defaults
    raise NotImplementedError


T = TypeVar('T')

OmitIfNone = Annotated[T, pydantic.WrapSerializer(omit_if_none)]
OmitIfDefault = Annotated[T, pydantic.WrapSerializer(omit_if_default)]


class Model(pydantic.BaseModel):
    a: OmitIfNone[int | None] = None
    b: OmitIfDefault[int | None] = None

print(Model(a=1, b=2).model_dump())
print(Model(a=None).model_dump())
```

wdyt?

**Reply by TristanSpeakEasy** [reply]
Just following up to see if there has been anymore progress around this, we are currently trying to map the relationship of optional and nullable fields in an OpenAPI document where these concepts can be set independently so for a particular Model you may have some fields that are optional and if not set shouldn't be present in the output, if their optional and nullable then if they aren't set they shouldn't be in the output but they should if explicitly set to `None` and if just Nullable and not optional then an optional field should always be present as `null` in the output.

And I'm struggling to get this to work with the current tools available

**Reply by joshorr** [reply]
I would also love to see this feature, ie: some way to omit a value when serializing a field/custom-type.

At the moment, the only way I found that really works is deleting the attributes off of the model to prevent the field from being serialized. This works, but it's less than ideal because I want the user to still be able to access these attributes.  Right now I implement a custom `__getattr__` method to return a sentinel object that indicates an attribute is missing.

I tried using a custom model serializer to pop/delete dicts keys after calling the `wrap` function. I got it to work fine for `model_dump()` but not `model_dump_json()` because I get this error: `error serializing to JSON: PydanticSerializationError: Unable to serialize unknown type` on my sentinel object when I call the `wrap(self)` function in the serializer while doing a `model_dump_json()` on the model object.  I could of course serialize the setinal object with some sort of special string or dictionary; but I don't want to do that because it takes a legal json value and treats it as 'missing', seems dangerous to do it that way.


If I could simply raise a `PydanticOmit` while Pydantic calls my serializer for the special sentinel object I am using to indicate value is currently missing, then it would solve this in a simple way. As it would cause Pydantic to not serialize the sentinel object at all, which is exactly what I want.

As a side-note: I also encountered a weird bug when using the `validate_assignment=True` option.  When setting the value, and if you raise a `PydanticOmit` while that value is validating via `core_schema.with_info_after_validator_function` on the value's class/type it will cause an error inside of Pydantic.  The validator works fine in all other situations, it only has the error while validating during assignment.

**Reply by ManuelAngel99** [reply]
Hi @adriangb @dmontagu is there any way to implement this behavior without manually deleting fields from the model dictionary at the moment?


We would also love to have this included so we could simplify scenarios like the following:

We have a class MyClass that has a query parameter that can be either a DefaultQuery or a FrontendProvidedQuery. If the FrontendProvidedQuery is present when parsing a json we want to deserialize it and store it in the query field, we also want to serialize it when calling model_dump() or model_dump_json() on instances of MyClass. If the default query is not present, we want to set query to a default value and skip it when serializing. 


In this case, the `OmitIfDefault` type annotation would come really handy, as it would allow us to have our code simplified to something like this:


```python
class MyClass(BaseModel):
    query: OmitIfDefault[DefaultQuery | FrontendProvidedQuery] = DefaultQuery(name='dummy')
```

Rather than having to manually implement our custom serializer which inspects all fields in the model and excludes the ones that are of type `DefaultQuery` (which is also way slower)

**Reply by joshorr** [reply]
@ManuelAngel99 I ended up writing a little library that marks fields as missing when the were not present (on pypi as [`pydantic-partials`](https://pypi.org/project/pydantic-partials/)).  If the field is missing, it will prevent it from being serialized (and it will be assigned a `Missing` sentenial value). The library can auto-apply the ability for missing to all fields (both inherited and on the current class); or you can manually do it by disabling the automatic mode and individually decide which specific fields can be missing via the `Partial[...]` annotation.

**Reply by Sam1320** [reply]
The `OmitIfDefault` would be extremely helpful for my use case. I couldn't find an existing issue or PR, Is this being worked on already or should I create a feature request/PR? @adriangb

**Reply by adriangb** [reply]
@Sam1320 could you just copy the 20-30 LOC above into your project?

**Reply by Sam1320** [reply]
> @Sam1320 could you just copy the 20-30 LOC above into your project?

ok 😢

**Reply by joshorr** [reply]
@adriangb The `PydanticOmit` does not work from a serializer, the above lines of code don't work.  See my comment above (https://github.com/pydantic/pydantic/discussions/5461#discussioncomment-9716405).  

I'm currently working around this by deleting the attributes on the model.  But I can't do that in every case, ie: `@computed_field`. I can't delete them but they may want to return my special value I am using to indicate something is "Missing" (I have someone asking me about it here: https://github.com/joshorr/pydantic-partials/issues/28).

I did try to override the `model_dump` but then that does not affect the `model_dump_json` output, and that presents a problem because there is no real special value I can inject into the `model_dump_json` for this because all values are valid json values.  So I then decided to delete the attributes on the model instead as a way forward.

I could potentially provide a full-model serializer function in my partial-base-class, but the parent-version is not called if a child provides its own full-model serializer function (ie: it overrides the parent version).  Just seems very messy to try and go that route due to this. And I also think I encountered some other limitation with full model serializes when doing a `model_dump_json` and trying to delete fields via a sentinel value, I can't remember the details right now.

If there was just a simple way to tell Pydantic to skip a field during serialization (ie: if a `PydanticOmit` is raised), that would be great!  I could just add raising`PydanticOmit`  to my `Missing` sentinel value class serializer (via the `__get_pydantic_core_schema__` method), then it would always be omitted during serialization that way in a much simpler manner.

**Reply by adriangb** [reply]
@sydney-runkle any reason we can't get `PydanticOmit` to work in serializers?

**Reply by sydney-runkle** [reply]
@adriangb,

I think it's doable, just not super high priority at the moment.

**Reply by joshorr** [reply]
@adriangb @sydney-runkle I know this is probably a pretty minor feature.  It's good to know that it might happen someday, so I can just look forward to it. For now I'll work around it the best I can.  Thank you both for your time!

**Reply by ManuelAngel99** [reply]
Would be amazing to have the PydanticOmit working @adriangb @sydney-runkle

**Reply by sydney-runkle** [reply]
Hey folks! This should help: https://github.com/pydantic/pydantic-core/pull/1535

**Reply by joshorr** [reply]
Thank you so much! I look forward to trying it out when it's released 😃

For my use case, I need to be able to evaluate this on every field. I suppose I could add the option to every field of the model.  And then if a field already had a `exclude_if` configured on it, I could call their function first and check its return result to see if it's already being excluded before running my own logic.

**Comment by larashores** [comment]
I have a very similar problem to this. I have a model of this form:

```python
class SuperModel(pydantic.BaseModel):
    field: int | None = None
    submodel: SubModel

class SubModel(pydantic.BaseModel):
    field: int | None = None
```

When serializing this, I want to always include the fields in `SubModel` that are set to `None`. For example, I would like to achieve the following:

```python
m = SuperModel(submodel=SubModel())
print(m.model_dump_json(exclude_none=True))
# > {"submodel": {"field": null}}
```

Basically, I want to ignore the `exclude_none` field for a specific model. I can achieve this with the following:
```python
@pydantic.model_serializer(mode="wrap")
def always_include_none(self, wrap: Callable[[Self], dict[str, Any]], info: pydantic.SerializationInfo) -> dict[str, Any]:
    values = wrap(self)
    for key, value in self.model_fields.items():
        if info.exclude and key in info.exclude:
            continue
        if info.include and key not in info.include:
            continue
        if key not in values and getattr(self, key, None) is None:
            values[value.alias if info.by_alias else key] = None
    return values
```

It would be nice if this was a builtin feature. Perhaps it can be an override in the config dict?
```python
class SubModel(pydantic.BaseModel):
    config = pydantic.ConfigDict(exclude_none=False)
```

Alternatively, maybe it can be a property of the field?
```python
class SuperModel(pydantic.BaseModel):
    submodel: Annotated[SubModel, pydantic.Field(exclude_none=False)]
```

**Comment by mikenerone** [comment]
This feels like something that should be done with `typing.Annotated`. However, it also also occurs to me that an even simpler approach might be to define a sentinel value (`pydantic.EXCLUDE`) and allow any field serializer to return that to indicate that the field should be excluded. This makes it trivial to exclude via a `@field_serializer` method while also allowing existing annotations like `WrapSerializer` to be leveraged on a per-type basis.

Edit: I thought about it a bit more and realized this doesn't actually lend itself to `Annotated` immediately, because Pydantic's built-in serializer logic (where the sentinel value might be noticed) runs _before_ the custom serializer function. There would need to be a little more code work to support it.

**Comment by KalleDK** [comment]
I have the same problem with an API where 2-3 fields should not be part of the json if they are None, but the 15 other fields should be included

**Comment by hampuskraft** [comment]
Inspired by previous responses, I came up with this solution, which works well enough for me at the moment—but I'd love to see a more ergonomic solution similar to what Adrian proposed.

```py
@dataclass
class OmitIfNone:
    pass

class AppResponseModel(BaseModel):
    @model_serializer
    def _serialize(self):
        omit_if_none_fields = {
            k
            for k, v in self.model_fields.items()
            if any(isinstance(m, OmitIfNone) for m in v.metadata)
        }

        return {k: v for k, v in self if k not in omit_if_none_fields or v is not None}

class UserResponseModel(AppResponseModel):
    id: int
    deleted: Annotated[Literal[True] | None, OmitIfNone()]
```

**Reply by mikenerone** [reply]
I did something similar, though I also needed exclusions based on some more complex criteria. I made an `ExcludeIf` annotation that lets you pass in a predicate (like `ExludeIf(lambda v: v < 5)` (it also supports a `when_used` argument that works like Pydantic's), then a few subclasses of it where the predicates are predefined on the class (`ExcludeIfNone()`, `ExcludeIfFalsy()`, `ExcludeIfEqual("a_value")`, `ExcludeIfIn(["some", "list"])`).

**Reply by ashb** [reply]
This approach works for the serialization, but has the unfortunate side effect that the schema the the returned type is empty now.

**Reply by joshorr** [reply]
I've been experimenting with something like this as well.  But there is a bit of a complication when you have multiple model serializers involved.

I wanted to create a very basic, library-style class model that would have this behavior generalized across all sorts of different kinds of models. This means some of these subclasses might want to further customize their model serialization behavior.

But if you set a model serializer on a subclass, the parent serializer does not get called automatically.  You can call it via `super().{parent-serializer-method-name}(...)` but you have to ensure you configure your subclass model serializer the same way, and pass the same arguments and so on (like the 'wrap' argument for wrap mode).  If the model changes how it wants to serialize things, the subclasses all have to be changed.  

It would be nice if when calling the default serialization via `mode='wrap'` that it would call the superclasses model serializer for you, and then the subclass does not have to worry about how/if it should call the super classes serializer. The doc says it serializes it the same way it normally would when using 'wrap' mode, but it does skip the superclasses model serializer function unfortunately.

**Reply by heyglen** [reply]
Thanks for this, it worked for me though I had to tweak it to take into account [Field aliases](https://docs.pydantic.dev/latest/concepts/fields/#field-aliases) and to run [PlainSerializer(s)](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.PlainSerializer):

```python
from dataclasses import dataclass

from pydantic import BaseModel, PlainSerializer, model_serializer

@dataclass
class OmitIfNone:
    pass

class AppResponseModel(BaseModel):
    @model_serializer
    def _serialize(self):
        skip_if_none = set()
        serialize_aliases = dict()

        # Gather fields that should omit if None
        for name, field_info in self.model_fields.items():
            if any(
                isinstance(metadata, OmitIfNone) for metadata in field_info.metadata
            ):
                skip_if_none.add(name)
            elif field_info.serialization_alias:
                serialize_aliases[name] = field_info.serialization_alias

        serialized = dict()

        for name, value in self:
            # Skip serializing None if it was marked with "OmitIfNone"
            if value is None and name in skip_if_none:
                continue
            serialize_key = serialize_aliases.get(name, name)

            # Run Annotated PlainSerializer
            for metadata in self.model_fields[name].metadata:
                if isinstance(metadata, PlainSerializer):
                    value = metadata.func(value)

            serialized[serialize_key] = value

        return serialized
```

**Comment by Viicos** [comment]
I'll try to see if this can be solved by https://github.com/pydantic/pydantic/issues/12056 and https://github.com/pydantic/pydantic/pull/10961.

**Comment by Viicos** [comment]
Thanks to @andresliszt, this should now be possible with https://github.com/pydantic/pydantic/pull/12141, that will land in the upcoming 2.12 version:

```python
from pydantic import BaseModel, Field

class Model(BaseModel):
    a: int | None = Field(exclude_if=lambda v: v is None)

Model(a=None).model_dump()
# {}
```

And with the proposal in https://github.com/pydantic/pydantic/issues/12056, it could be made even more performant by not leaving `pydantic-core` at all:

```python

class Model(BaseModel):
    a: int | None = Field(exclude_if=IsNone())
```

Users may also be interested in the experimental [`MISSING` sentinel](https://docs.pydantic.dev/dev/concepts/experimental/#missing-sentinel) (also introduced in 2.12), applying automatic field exclusion.

Regarding setting `exclude_none` in model configuration (https://github.com/pydantic/pydantic/discussions/5461#discussioncomment-7180952), this is tracked in https://github.com/pydantic/pydantic/issues/10141.

**Reply by bluenote10** [reply]
It should be mentioned that `exclude_if` doesn't fully solve the problem yet, because it doesn't have proper support for `'serialization'` mode JSON schemas.

What's happening currently: The field will appear in `"required"` in the schema. For output/serialization mode, the `"required"` block is semantically rather a _guaranteed_ field list. Including the field there is incorrect, exactly because the `exclude_if` makes the inclusion _conditional_, i.e., it is exactly no longer _guaranteed_.

This is a bit of a showstopper for use in FastAPI, because the OpenAPI specs still do not properly express the behavior (which in turns breaks auto-generated clients).

This is tracked in: https://github.com/pydantic/pydantic/issues/12424
